extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -554,59 +568,52 @@ class LifecycleTask extends BackbeatTask {
-                    const req = this.s3target.headObject({
-                        Bucket: bucketData.target.bucket,
-                        Key: version.Key,
-                        VersionId: version.VersionId,
-                    });
-                    attachReqUids(req, log);
-                    return req.send((err, data) => {
-                        if (err) {
-                            log.error('failed to get object', {
-                                method: 'LifecycleTask._compareVersions',
-                                error: err,
-                                bucket: bucketData.target.bucket,
-                                objectKey: version.Key,
-                                versionId: version.VersionId,
-                            });
-                            return next(err);
-                        }
-                        const objDate = new Date(data.LastModified);
-                        const now = Date.now();
-                        // time diff in ms
-                        const diff = now - objDate;
-                        const daysSinceInitiated =
-                            Math.floor(diff / (1000 * 60 * 60 * 24));
-                        const ncve = 'NoncurrentVersionExpiration';
-                        const ncd = 'NoncurrentDays';
-                        if (applicableRules[ncve] &&
-                        applicableRules[ncve][ncd] &&
-                        daysSinceInitiated >= applicableRules[ncve][ncd]) {
-                            const entry = {
-                                action: 'deleteObject',
-                                target: {
-                                    owner: bucketData.target.owner,
-                                    bucket: bucketData.target.bucket,
-                                    key: version.Key,
-                                    version: data.VersionId,
-                                },
-                            };
-                            this.sendObjectEntry(entry, err => {
-                                if (!err) {
-                                    log.debug('sent object entry for ' +
-                                    'consumption', {
-                                        method: 'LifecycleTask.' +
-                                            '_compareVersions',
-                                        entry,
-                                    });
-                                }
-                            });
+                        // `_compareObject` should not apply for delete markers
+                        if (version.StorageClass && version.Size &&
+                        version.ETag) {
+                            return this._compareObject(bucketData, version,
+                                applicableRules, log, next);
                         }
                         return next();
-                    });
+                    }
+
+                    const now = Date.now();
```

Comment:
May be better organized putting this code into a separate method like `applyNoncurrentRules()`

Reply:
Sounds good to me 👍 
Possible chance of refactoring this logic as well since there are multiple areas where the dates are compared. Will need to check again to make sure

Reply:
Sounds good to factorize, just make sure there's enough test coverage in and around this function.